linera-core/src/value_cache.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A least-recently used cache of [`HashedCertificateValue`]s.
+
+#[cfg(test)]
+#[path = "unit_tests/value_cache_tests.rs"]
+mod unit_tests;
+
+use std::{borrow::Cow, collections::BTreeSet, num::NonZeroUsize};
+
+use linera_base::crypto::CryptoHash;
+use linera_chain::data_types::{Certificate, HashedCertificateValue, LiteCertificate};
+use lru::LruCache;
+use tokio::sync::Mutex;
+#[cfg(with_metrics)]
+use {
+    linera_base::{prometheus_util, sync::Lazy},
+    prometheus::IntCounterVec,
+};
+
+use crate::worker::WorkerError;
+
+/// The default cache size.
+const DEFAULT_VALUE_CACHE_SIZE: usize = 1000;
+
+/// A counter metric for the number of cache hits in the [`CertificateValueCache`].
+#[cfg(with_metrics)]
+static CACHE_HIT_COUNT: Lazy<IntCounterVec> = Lazy::new(|| {
+    prometheus_util::register_int_counter_vec(
+        "certificate_value_cache_hit",
+        "Cache hits in `CertificateValueCache`",
+        &[],
+    )
+    .expect("Counter creation should not fail")
+});
+
+/// A counter metric for the number of cache hits in the [`CertificateValueCache`].
+#[cfg(with_metrics)]
+static CACHE_MISS_COUNT: Lazy<IntCounterVec> = Lazy::new(|| {
+    prometheus_util::register_int_counter_vec(
+        "certificate_value_cache_miss",
+        "Cache misses in `CertificateValueCache`",
+        &[],
+    )
+    .expect("Counter creation should not fail")
+});
+
+/// A least-recently used cache of [`HashedCertificateValue`]s.
+pub struct CertificateValueCache {
+    cache: Mutex<LruCache<CryptoHash, HashedCertificateValue>>,
```

Comment:
Why not having the `Arc` here as well? The `Mutex` alone seems like an anti-pattern.

Reply:
I aimed to do something similar to what [`DashMap`](https://docs.rs/dashmap/latest/dashmap/struct.DashMap.html) does. My reasoning for that is:
- I'd like `CertificateValueCache` to be a thread-safe cache that can be shared, so it requires interior mutability (therefore using `Mutex` inside)
- One way to share the cache is using an `Arc`, but that's left as a choice to the user
- Making `Arc<CertificateValueCache>` also makes it clearer that `.clone()` does _not_ clone the cache contents

Let me know what you think, I can change it either way.